src/lib/salad/hll.c

```diff
@@ -0,0 +1,223 @@
+#include "hll.h"
+#include "hll_empirical.h"
+
+#include <trivia/util.h>
+
+#include <stddef.h>
+#include <math.h>
+#include <assert.h>
+
+/* 6 bits are needed to store the number of leading zeros of 64 bit hash */
+#define HLL_REGISTER_BITS  6
+
+static inline size_t
+hll_n_registers(uint8_t precision)
+{
+	return 1U << precision;
+}
+
+static inline int
+hll_is_valid_cache(double cached)
+{
+	return cached > 0;
+}
+
+static inline void
+hll_invalidate_cache(struct hll *hll)
+{
+	hll->cached_estimation = -1.f;
+}
+
+struct hll *
+hll_create(uint8_t precision)
+{
+	assert(precision >= HLL_MIN_PRECISION);
+	assert(precision <= HLL_MAX_PRECISION);
+	struct hll *hll = xcalloc(1, sizeof(*hll));
+	if (hll) {
```

Comment:
hll cannot be `NULL` (please look what `xalloc` does).

Reply:
Thank you for noticing. I forgot to delete it after using malloc.